control-plane/connect-inject/controllers/pod/pod_controller_test.go

```diff
@@ -322,20 +298,519 @@ func TestWorkloadWrite(t *testing.T) {
+				Output:      constants.KubernetesSuccessReasonMsg,
+				Description: constants.ConsulKubernetesCheckName,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			run(t, tc)
+		})
+	}
+}
+
+func TestProxyConfigurationWrite(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		name                       string
+		pod                        *corev1.Pod
+		podModifier                func(pod *corev1.Pod)
+		expectedProxyConfiguration *pbmesh.ProxyConfiguration
+
+		tproxy          bool
+		overwriteProbes bool
+		metrics         bool
+		telemetry       bool
+	}
+
+	run := func(t *testing.T, tc testCase) {
+		ns := corev1.Namespace{ObjectMeta: metav1.ObjectMeta{
+			Name: metav1.NamespaceDefault,
+		}}
+
+		nsTproxy := corev1.Namespace{ObjectMeta: metav1.ObjectMeta{
+			Name: "tproxy-party",
+			Labels: map[string]string{
+				constants.KeyTransparentProxy: "true",
+			},
+		}}
+
+		if tc.podModifier != nil {
+			tc.podModifier(tc.pod)
+		}
+
+		fakeClient := fake.NewClientBuilder().WithRuntimeObjects(&ns, &nsTproxy).Build()
+
+		// Create test consulServer server.
+		testClient := test.TestServerWithMockConnMgrWatcher(t, func(c *testutil.TestServerConfig) {
+			c.Experiments = []string{"resource-apis"}
+		})
+		resourceClient, err := consul.NewResourceServiceClient(testClient.Watcher)
+		require.NoError(t, err)
+
+		// Create the pod controller.
+		pc := &Controller{
+			Client:                   fakeClient,
+			Log:                      logrtest.New(t),
+			ConsulClientConfig:       testClient.Cfg,
+			ConsulServerConnMgr:      testClient.Watcher,
+			AllowK8sNamespacesSet:    mapset.NewSetWith("*"),
+			DenyK8sNamespacesSet:     mapset.NewSetWith(),
+			EnableTransparentProxy:   tc.tproxy,
+			TProxyOverwriteProbes:    tc.overwriteProbes,
+			EnableTelemetryCollector: tc.telemetry,
+			ResourceClient:           resourceClient,
+		}
+
+		if tc.metrics {
+			pc.MetricsConfig = metrics.Config{
+				DefaultEnableMetrics:        true,
+				DefaultPrometheusScrapePort: "5678",
+			}
+		}
+
+		// Test writing the pod to a HealthStatus
+		err = pc.writeProxyConfiguration(context.Background(), *tc.pod)
+		require.NoError(t, err)
+
+		req := &pbresource.ReadRequest{
+			Id: getProxyConfigurationID(tc.pod.GetName(), metav1.NamespaceDefault, constants.DefaultConsulPartition),
+		}
+		actualRes, err := resourceClient.Read(context.Background(), req)
+
+		if tc.expectedProxyConfiguration == nil {
+			require.Error(t, err)
+			s, ok := status.FromError(err)
+			require.True(t, ok)
+			require.Equal(t, codes.NotFound, s.Code())
+			return
+		}
+
+		require.NoError(t, err)
+		require.NotNil(t, actualRes)
+
+		require.Equal(t, tc.pod.GetName(), actualRes.GetResource().GetId().GetName())
+		require.Equal(t, constants.DefaultConsulNS, actualRes.GetResource().GetId().GetTenancy().GetNamespace())
+		require.Equal(t, constants.DefaultConsulPartition, actualRes.GetResource().GetId().GetTenancy().GetPartition())
```

Comment:
~~ Could maybe add an `ID` equals helper for the tests that need it, just for better legibility